pkg/deps/latest.go

```diff
@@ -0,0 +1,244 @@
+
+			var newChild *node
+			if alreadyPicked {
+				// We already picked a component with the same component name. Check to
+				// make sure the version works with this component's dependency
+				// requirements
+				successfulPick, err := r.processPicked(dep, curNode, picked)
+				if err != nil {
+					return nil, err
+				}
+				if !successfulPick {
+					success = false
+					break
+				}
+				newChild = picked
+			} else {
+				// Do selection for the first time. We'll pick the latest component
+				// version of the component that satisfies the requirements.
+				latest, err := r.pickLatest(dep, curNode)
+				if err != nil {
+					return nil, err
+				}
+				if latest == nil {
+					success = false
+					break
+				}
+				newChild = latest
+			}
+			if !newChild.meta.visibleTo(curNode.meta) {
+				return nil, fmt.Errorf("component %v is not visible to component %v", newChild.meta, curNode.meta)
+			}
+			curNode.children = append(curNode.children, newChild)
+		}
+
+		if !success {
+			err := r.tryDowngrade(curNode)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return curNode.children, nil
+}
+
+// processPicked checks to make sure an already picked component works with the
+// existing constraints
+func (r *graphBuilder) processPicked(dep *requestedDep, curNode, picked *node) (bool, error) {
+	if picked.meta.version.GTE(dep.version) {
+		// It works!
+		return true, nil
+	}
+	if dep.version.GT(picked.maxOfMinReq) {
+		picked.maxOfMinReq = dep.version
+	}
+	// It doesn't work. Indicate to the caller that a downgrade needs to be performed.
+	return false, nil
+}
+
+// pickLatest picks the latest component for a given dependency requirement.
+func (r *graphBuilder) pickLatest(dep *requestedDep, curNode *node) (*node, error) {
+	sorted, ok := r.componentVersions[dep.componentName]
+	if !ok {
+		return nil, fmt.Errorf("could not find sorted component versions for component name %q", dep.componentName)
+	}
+	latest, err := sorted.latest(r.searchOpts)
+	if err != nil {
+		return nil, err
+	}
+	if latest.version.LT(dep.version) {
+		// This is a special error case: the latest version version of the dep
+		// doesn't work with the curNode
+		//
+		// If the latest version of the dependency is less than the requested
+		// version, our only option is to try to downgrade the parent node. We pass
+		// back nil to indicate as such.
+		return nil, nil
+	}
+	newChild := &node{
+		meta:        latest,
+		maxOfMinReq: dep.version,
+	}
+	return newChild, nil
+}
+
+// tryDowngrade tries to downgrade the current node version, in-place.
+//
+// This is done in the case where there was some, possibly recoverable
+// incomptibility. Because we always select the latest version of dependencies,
+// we don't have the option to upgrade the picked version. Instead, we can try
+// to downgrade current node and start the dependency process over for the
+// current node.
+func (r *graphBuilder) tryDowngrade(curNode *node) error {
+	if curNode.fixed {
```

Comment:
It's unobvious within this function what the affect of setting currNode.children to null and meta to previous would do. Those are not concepts that actually are from this function.  Instead yoru depending upon class level data and how this code was called. 

Consider returning data instead.  TryGetLowerVersion which returns  lower version if possible.  Don't mutate anything in here, and only modify the data structures from the calling piece of code.  

Notice your comments " Clear out the children and try again".  That's not what the code does.  That's only what the code does in the context of the other code that uses this code has the affect of doing.   I encourage you to apply the mindset of returning data and only mutating it as close to the actual relevant data structure manipulating code as much as possible.